src/jobserver.cc

```diff
@@ -0,0 +1,208 @@
+
+uint8_t Jobserver::Slot::GetExplicitValue() const {
+  assert(IsExplicit());
+  return static_cast<uint8_t>(value_);
+}
+
+bool Jobserver::ParseMakeFlagsValue(const char* makeflags_env,
+                                    Jobserver::Config* config,
+                                    std::string* error) {
+  *config = Config();
+
+  if (!makeflags_env || !makeflags_env[0]) {
+    /// Return default Config instance with kModeNone if input is null or empty.
+    return true;
+  }
+
+  // Decompose input into vector of space or tab separated string pieces.
+  std::vector<StringPiece> args;
+  const char* p = makeflags_env;
+  while (*p) {
+    const char* next_space = strpbrk(p, " \t");
+    if (!next_space) {
+      args.emplace_back(p);
+      break;
+    }
+
+    if (next_space > p)
+      args.emplace_back(p, next_space - p);
+
+    p = next_space + 1;
+  }
+
+  // clang-format off
+  //
+  // From:
+  // https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html
+  //
+  // """
+  // Your tool may also examine the first word of the MAKEFLAGS variable and
+  // look for the character n. If this character is present then make was
+  // invoked with the ‘-n’ option and your tool may want to stop without
+  // performing any operations.
+  // """
+  //
+  // Where according to
+  // https://www.gnu.org/software/make/manual/html_node/Options_002fRecursion.html
+  // MAKEFLAGS begins with all "flag letters" passed to make.
+  //
+  // Experimentation shows that GNU Make 4.3, at least, will set MAKEFLAGS with
+  // an initial space if no letter flag are passed to its invocation (except -j),
+  // i.e.:
+  //
+  //    make -ks --> MAKEFLAGS="ks"
+  //    make -j  --> MAKEFLAGS=" -j"
+  //    make -ksj --> MAKEFLAGS="ks -j"
+  //    make -ks -j3  --> MAKEFLAGS="ks -j3 --jobserver-auth=3,4"
+  //    make -j3      --> MAKEFLAGS=" -j3 --jobserver-auth=3,4"
+  //
+  // However, other jobserver implementation will not, for example the one
+  // at https://github.com/rust-lang/jobserver-rs will set MAKEFLAGS to just
+  // "--jobserver-fds=R,W --jobserver-auth=R,W" instead, without an initial
+  // space.
+  //
+  // Another implementation is from Rust's Cargo itself which will set it to
+  // "-j --jobserver-fds=R,W --jobserver-auth=R,W".
+  //
+  // For the record --jobserver-fds=R,W is an old undocumented and deprecated
+  // version of --jobserver-auth=R,W that was implemented by GNU Make before 4.2
+  // was released, and some tooling may depend on it. Hence it makes sense to
+  // define both --jobserver-fds and --jobserver-auth at the same time, since
+  // the last recognized one should win in client code.
+  //
+  // The initial space will have been stripped by the loop above, but we can
+  // still support the requirement by ignoring the first arg if it begins with a
+  // dash (-).
+  //
+  // clang-format on
+  if (!args.empty() && args[0][0] != '-' &&
+      !!memchr(args[0].str_, 'n', args[0].len_)) {
```

Comment:
why the `!!`?

Reply:
It is just a short way to force a value to a bool, I have replaced it with `memchr(...) != nullptr` for clarity.